File: pp-exercises/mpi/exercise3/ring-datatype-operator.c
# include <stdio.h>
# include <mpi.h>

typedef struct myStruct { double x; int i; } myStruct;
MPI_Datatype myType;      // MPI datatype for myStruct
MPI_User_function myFun;  // reduction operation for myType
MPI_Op myOp;              // MPI operator for myFun

int main(int argc, char *argv[])  // user defined datatype and operator
{
    myStruct s, res;
    int rank, len[2];
    MPI_Aint disp[2];
    MPI_Datatype type[2];
 
    MPI_Init(&argc, &argv);

    len[0] = 1;
    len[1] = 1;

    MPI_Get_address(&s.x, &disp[0]);
    MPI_Get_address(&s.i, &disp[1]);
 
    disp[1] -= disp[0];
    disp[0] -= disp[0];

    type[0] = MPI_DOUBLE;
    type[1] = MPI_INT;

    MPI_Type_create_struct(2, len, disp, type, &myType);
    MPI_Type_commit(&myType);
    MPI_Op_create(myFun, 1, &myOp);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    s.x = rank;
    s.i = rank;

    MPI_Reduce(&s, &res, 1, myType, myOp, 0, MPI_COMM_WORLD);

    if (rank == 0) printf("result: %f, %i\n", res.x, res.i);

    MPI_Type_free(&myType);
    MPI_Op_free(&myOp);
    MPI_Finalize();
    return 0;
}

void myFun(void* in, void* inout, int* len, MPI_Datatype *type)
{
    myStruct *i = (myStruct *) in;
    myStruct *io = (myStruct *) inout;
    int j;

    if (*type == myType) {
	for (j = 0; j < *len; j++) {
	    io[j].x += i[j].x;
	    io[j].i += i[j].i;
	}
    } else {
	fputs("myFun: unknown type\n", stderr);
	MPI_Abort(MPI_COMM_WORLD, 1);
    }
}

File: pp-exercises/mpi/exercise3/ring-datatype-operator.f90
module module_my_struct

  type my_struct
     real(8) :: x
     integer :: i
  end type my_struct
  
  integer, save :: my_type  ! MPI datatype for type(my_struct)
  external      :: my_fun   ! reduction operation for my_type
  integer, save :: my_op    ! MPI operator for my_fun
end 


program main  ! user defined datatype and operator

  use module_my_struct
  implicit none
  include 'mpif.h'
  type(my_struct) :: s, res
  integer :: rank, len(2), type(2), ierror
  integer(MPI_ADDRESS_KIND) :: disp(2)

  call mpi_init(ierror)

  len(1) = 1
  len(2) = 1

  call mpi_get_address(s%x, disp(1), ierror)
  call mpi_get_address(s%i, disp(2), ierror)
 
  disp(2) = disp(2) - disp(1)
  disp(1) = disp(1) - disp(1)

  type(1) = MPI_REAL8
  type(2) = MPI_INTEGER

  call mpi_type_create_struct(2, len, disp, type, my_type, ierror)
  call mpi_type_commit(my_type, ierror)
  call mpi_op_create(my_fun, .true., my_op, ierror)
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierror)

  s%x = rank
  s%i = rank

  call mpi_reduce(s, res, 1, my_type, my_op, 0, MPI_COMM_WORLD, ierror)

  if (rank == 0) write(6,*) "result: ", res%x, res%i

  call mpi_type_free(my_type, ierror)
  call mpi_op_free(my_op, ierror)
  call mpi_finalize(ierror)
end


subroutine my_fun(in, inout, len, type)

  use module_my_struct, exclude => my_fun
  implicit none
  include 'mpif.h'
  type(my_struct), intent(in) :: in(*)
  type(my_struct), intent(inout) :: inout(*)
  integer, intent(in) :: len, type
  integer :: i, ierror

  if (type == my_type) then   
     do i = 1, len
        inout(i)%x = inout(i)%x + in(i)%x
        inout(i)%i = inout(i)%i + in(i)%i
     enddo
  else
     write(0,*) "my_fun: unknown type"
     call mpi_abort(MPI_COMM_WORLD, 1, ierror)
  endif
end
